File: 03-week-3--additional-practice/02-bonus-3--rotate-array-clockwise/solutions/rotate_array.js
function rotateArray(arr, k) {
  const rotations = k % arr.length;
  const removed = arr.splice(arr.length - rotations, rotations);

  return removed.concat(arr);
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: [4, 1, 2, 3]");
  console.log(rotateArray([1, 2, 3, 4], 1));

  console.log("");

  console.log("Expecting: [2, 3, 1]");
  console.log(rotateArray([1, 2, 3], 2));

  console.log("");

  console.log("Expecting: [1, 2, 3]");
  console.log(rotateArray([1, 2, 3], 3));

  console.log("");

  console.log("Expecting: [1, 2, 3]");
  console.log(rotateArray([1, 2, 3], 0));

  console.log("");

  console.log("Expecting: [2, 3, 1]");
  console.log(rotateArray([1, 2, 3], 5));

  console.log("");

  console.log("Expecting: []");
  console.log(rotateArray([], 7));
}

module.exports = rotateArray;

// Please add your pseudocode to this file
/***********************************************************************************
 * initialize rotations to store the remainder of k / array length
 * remove rotations number elements from end of array
 * return removed elements + remaining elements
 * *********************************************************************************/

// And a written explanation of your solution
/***********************************************************************************
 * We can solve this problem by figuring out how many elements to remove from the end
 * of the array, and then adding those removed elements to the front of the array. To
 * do this, we need to account for when k is the same or larger than the array's length.
 * This is ascertained by getting the remainder of k / array length.
 * **********************************************************************************/

File: 03-week-3--additional-practice/02-bonus-3--rotate-array-clockwise/solutions/rotate_array.rb
def rotate_array(arr, k)
  # avoid divide by zero error in Ruby, which will otherwise occur on L5
  return arr if arr.empty? || k.zero?

  rotations = k % arr.length
  removed = arr.slice!(arr.length - rotations, rotations)
  removed + arr
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: [4, 1, 2, 3]"
  print rotate_array([1, 2, 3, 4], 1)

  puts

  puts "Expecting: [2, 3, 1]"
  print rotate_array([1, 2, 3], 2)

  puts

  puts "Expecting: [1, 2, 3]"
  print rotate_array([1, 2, 3], 3)
  # Don't forget to add your own!

  puts

  puts "Expecting: [1, 2, 3]"
  print rotate_array([1, 2, 3], 0)

  puts

  puts "Expecting: [2, 3, 1]"
  print rotate_array([1, 2, 3], 5)

  puts

  puts "Expecting: []"
  print rotate_array([], 7)
end

# Please add your pseudocode to this file
##########################################################################################
 # initialize rotations to store the remainder of k / array length
 # remove rotations number elements from end of array
 # return removed elements + remaining elements
##########################################################################################

# And a written explanation of your solution
##########################################################################################
 # We can solve this problem by figuring out how many elements to remove from the end
 # of the array, and then adding those removed elements to the front of the array. To
 # do this, we need to account for when k is the same or larger than the array's length.
 # This is ascertained by getting the remainder of k / array length.
##########################################################################################